### PR TITLE
Token become immutable, performance optimizations

### PR DESCRIPTION
**Pull request by keradus**
TODO:
- [x] rebase on master

```
Executing on Symfony 2.7 with 45files that reports violations

before
> Checked all files in 77.743 seconds, 20.000 MB memory used
> Checked all files in 78.761 seconds, 20.000 MB memory used

after improvement
> Checked all files in 70.417 seconds, 20.000 MB memory used
> Checked all files in 71.144 seconds, 20.000 MB memory used

About 10% improvement
```

Deps:
- [x] #2819
- [x] #2816
- [x] #2817
- [x] #2823

### CODE REVIEW

File: src/Tokenizer/Tokens.php [hunk 1]
```diff
@@ -1085,18 +1175,26 @@ public function isMonolithicPhp()
...
+
+            return true;
         }
 
-        return true;
+        if ($this->isTokenKindFound(T_INLINE_HTML)) {
+            return false;
+        }
+
+        return 1 >= ($this->countTokenKind(T_OPEN_TAG) + $this->countTokenKind(T_OPEN_TAG_WITH_ECHO));
```
Comment (julienfalque) [inline src/Tokenizer/Tokens.php]: Yoda style in expressions using `<` or `>` is confusing IMO, I would swap the operands.

File: src/Tokenizer/Tokens.php [hunk 2]
```diff
@@ -1218,11 +1314,41 @@ private function changeCodeHash($codeHash)
...
+     */
+    private function extractTokenKind($token)
+    {
+        return $token instanceof Token
             ? ($token->isArray() ? $token->getId() : $token->getContent())
             : (is_array($token) ? $token[0] : $token)
-        ;
-
-        $this->foundTokenKinds[$tokenKind] = true;
+            ;
```
Comment (julienfalque) [inline src/Tokenizer/Tokens.php]: This semicolon should be less indented.

File: src/Tokenizer/Tokens.php [hunk 3]
```diff
@@ -874,9 +945,14 @@ public function clearAt($index)
      *
      * @param int                $index
      * @param Token|array|string $token token prototype
+     *
+     * @deprecated since 2.4, use offsetSet instead
      */
     public function overrideAt($index, $token)
     {
+        @trigger_error(__METHOD__.' is deprecated and will be removed in 3.0, use offsetSet instead.', E_USER_DEPRECATED);
+        self::$isLegacyMode = true;
```
Comment (keradus) [inline src/Tokenizer/Tokens.php]: no, it would change meaning of code.

File: src/Tokenizer/Tokens.php (hunk 1 above)
Comment (keradus) [inline src/Tokenizer/Tokens.php]: we use yoda across the project in such cases nevertheless

File: src/Tokenizer/Tokens.php (hunk 2 above)
Comment (keradus) [inline src/Tokenizer/Tokens.php]: done

File: src/Tokenizer/Tokens.php [hunk 4]
```diff
@@ -68,11 +68,20 @@ class Tokens extends \SplFixedArray
...
      *
-     * @var bool[]
+     * @var array<string|int, int>
      */
     private $foundTokenKinds = array();
 
     /**
+     * @var bool
+     *
+     * @internal
```
Comment (keradus) [inline src/Tokenizer/Tokens.php]: done

File: src/Tokenizer/Token.php [hunk 5]
```diff
@@ -61,6 +61,10 @@ public function __construct($token)
             $this->isArray = true;
             $this->id = $token[0];
             $this->content = $token[1];
+
+            if ($token[0] && '' === $token[1]) {
+                throw new \InvalidArgumentException('Cannot set empty content for id-based Token.');
```
Comment (keradus) [inline src/Tokenizer/Token.php]: missing validation

File: src/Tokenizer/Token.php [hunk 6]
```diff
@@ -495,6 +517,7 @@ public function setContent($content)
 
         // setting empty content is clearing the token
         if ('' === $content) {
+            @trigger_error(__METHOD__.' shall not be use to clear token, use Tokens::clearAt instead.', E_USER_DEPRECATED);
```
Comment (keradus) [inline src/Tokenizer/Token.php]: > and will throw an exception in 3.0

it won't, as the method won't be there at all

Comment (julienfalque) [inline src/Tokenizer/Token.php]: Then isn't this deprecation notice useless as we already have the one for the method deprecation itself?

Comment (keradus) [inline src/Tokenizer/Token.php]: not fully, as when set with empty content, we raise even bigger issue to dev

**Comment by keradus** [comment]
> Please add tests that cover `Tokens` changes regarding "immutable mode".

https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2818/commits/9b63057380de9a557b7bbc8affcca64a8bc1b92e

File: src/Tokenizer/Token.php (hunk 6 above)
Comment (julienfalque) [inline src/Tokenizer/Token.php]: 👍 but please fix the typo: _shall not be use**d**_.

Comment (keradus) [inline src/Tokenizer/Token.php]: done

**Pull request closed without merging**